Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp

```diff
@@ -1636,24 +1675,68 @@ void MediaPlayerPrivateGStreamer::processBufferingStats(GstMessage* message)
     // The Nexus playpump buffers a lot of data. Let's add it as if it had been buffered by the GstQueue2
     // (only when using in-memory buffering), so we get more realistic percentages.
     if (!m_downloadBuffer && m_vidfilter) {
+        int originalBufferingPercentage = m_bufferingPercentage;
+        GST_DEBUG("[Buffering] buffering percentage from message: %d%%", originalBufferingPercentage);
+        int updatedBufferingPercentage = m_bufferingPercentage;
+        int correctedBufferingPercentage = m_bufferingPercentage;
+
         GstObject *queue2 = GST_MESSAGE_SRC(message);
         guint maxSizeBytes = 0;
         // Current-level-bytes seems to be inacurate, so we compute its value from the buffering percentage
         g_object_get(queue2, "max-size-bytes", &maxSizeBytes, nullptr);
+        GST_DEBUG("[Buffering] max size bytes of queue2: %d", maxSizeBytes);
+
+        if (m_bufferingPercentage == 0) {
+            guint currentLevelBytes = 0;
+            g_object_get(queue2, "current-level-bytes", &currentLevelBytes, nullptr);
+            GST_DEBUG("[Buffering] current bytes level of queue2: %d", currentLevelBytes);
+
+            updatedBufferingPercentage = currentLevelBytes > maxSizeBytes ? 100 : (int)(currentLevelBytes * 100 / maxSizeBytes);
+            m_bufferingPercentage = updatedBufferingPercentage;
+        }
 
         guint playpumpBufferedBytes = 0;
         g_object_get(GST_OBJECT(m_vidfilter.get()), "buffered-bytes", &playpumpBufferedBytes, nullptr);
+        GST_DEBUG("[Buffering] buffered bytes by playpump: %d", playpumpBufferedBytes);
+
+        guint multiqueueBufferedBytes = 0;
+        if (m_multiqueue) {
+            GUniqueOutPtr<GstStructure> stats;
+            g_object_get(m_multiqueue.get(), "stats", &stats.outPtr(), nullptr);
+            const GValue* queues = gst_structure_get_value(stats.get(), "queues");
+            guint size = gst_value_array_get_size(queues);
+            for (guint i = 0; i < size; i++) {
+                guint bytes = 0;
+                if (gst_structure_get_uint(gst_value_get_structure(gst_value_array_get_value(queues, i)), "bytes", &bytes))
+                    multiqueueBufferedBytes += bytes;
+            }
```

Comment:
you're counting both audio and video data here (as the same multiqueue handles both), but later you're comparing it to max-size-bytes of only one queue2 to calculate percentages- is this correct?

Reply:
yes, because queue2 process data which contains A/V, before qtdemux the A/V are kept as one bytes stream